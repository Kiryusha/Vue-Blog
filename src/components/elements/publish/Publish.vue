<template lang="pug">
  section.publish(:class="{'_loaded': loaded}")
    .title: h1 Отправить новость
    .content
      form.form(
        @submit.prevent=""
      )
        .row
          .row-title Название
          Textarea(
            :error="getErrorMessage($v.title)"
            :valid="!$v.title.$invalid"
            @input="$v.title.$touch()"
            v-model="title"
            :maxlength="140"
          )
        .row
          .row-title
            span Символьный код
            span.question
              Icon(
                :name="'question'"
                width=16
                height=16
                v-tooltip.click.top=`{
                  content: 'Адрес, который будет выводиться в url. Должен быть уникальным.'
                }`
              )
          Textarea(
            :error="getErrorMessage($v.code)"
            :valid="!$v.code.$invalid"
            @input="$v.code.$touch()"
            v-model="code"
            :maxlength="50"
          )
        .row
          .row-title Категория
          Textarea(
            :error="getErrorMessage($v.category)"
            :valid="!$v.category.$invalid"
            @input="$v.category.$touch()"
            v-model="category"
            :maxlength="50"
          )
        .row
          .row-title Путь к картинке для анонса
          Textarea(
            :rows=1
            v-model="previewPicture"
            :maxlength="1000"
          )
        .row
          .row-title Описание для анонса
          Textarea(
            :rows=3
            v-model="previewText"
            :maxlength="311"
          )
        .row
          .row-title
            span Детальное описание
            span.question
              Icon(
                :name="'question'"
                width=16
                height=16
                v-tooltip.click.top=`{
                  content: 'Можно использовать HTML.'
                }`
              )
          Textarea(
            :rows=5
            :error="getErrorMessage($v.detailText)"
            :valid="!$v.detailText.$invalid"
            @input="$v.detailText.$touch()"
            v-model="detailText"
            :maxlength="5000"
          )
        .row._controls
          Button(
            :view="'fz16'"
            @click="submit"
          ) Отправить
          Button(
            v-if="this.state === 'edit'"
            :view="'fz16'"
            @click="$router.push('/blog/')"
          ) Отмена
</template>

<script>
import { mapState, mapActions } from 'vuex';
import { required } from 'vuelidate/lib/validators';
import Textarea from 'Components/elements/general/Textarea';
import Button from 'Components/elements/general/Button';
import Icon from 'Components/elements/general/Icon';

export default {
  components: {
    Textarea,
    Button,
    Icon,
  },
  props: {
    postCode: {
      type: String,
      default: '',
    },
  },
  data() {
    return {
      id: '',
      title: '',
      code: '',
      category: '',
      previewPicture: '',
      previewText: '',
      detailText: '',
      submitted: false,
      loaded: false,
    };
  },
  computed: {
    ...mapState({
      post: state => state.post.data,
      userId: state => state.user.id,
      username: state => state.user.name,
    }),
    state() {
      return this.postCode && this.postCode.length ? 'edit' : 'post';
    },
  },
  created() {
    this.loaded = true;
    if (this.state === 'edit') {
      this.fetchData(this.postCode);
    }
    this.$Progress.finish();
  },
  watch: {
    state(type) {
      if (type === 'edit') {
        this.fetchData(this.postCode);
      }
    },
  },
  methods: {
    ...mapActions([
      'fetchPost',
      'sendPost',
      'updatePost',
    ]),
    getErrorMessage(field) {
      let message = '';

      if (field.$error && this.submitted) {
        if (field.checkTitle === false) {
          message = 'Только буквы, числа и пунктуация: ", . : ; ! ?"';
        }

        if (field.checkCode === false) {
          message = 'Только латиница и числа';
        }

        if (field.required === false) {
          message = 'Обязательное поле';
        }
      }

      return message;
    },
    async submit() {
      this.$v.$touch();
      this.submitted = true;

      if (!this.$v.$invalid) {
        const data = {
          title: this.title,
          id: this.id,
          code: this.code.toLowerCase(),
          category: this.category ? this.category : 'Без категории',
          previewText: this.previewText,
          previewPicture: this.previewPicture,
          detailText: this.detailText,
          username: this.username,
          userId: this.userId,
        };

        if (this.state === 'edit') {
          this.response = await this.updatePost(data);
        } else {
          this.response = await this.sendPost(data);
        }

        if (this.response) {
          this.$modal.show('response', { message: this.response.data.message });
        }

        this.submitted = false;
        this.title = '';
        this.code = '';
        this.category = '';
        this.previewPicture = '';
        this.previewText = '';
        this.detailText = '';

        this.$router.push('/blog/');
      }
    },
    async fetchData(code) {
      await this.fetchPost(code);

      this.id = this.post._id;
      this.title = this.post.title;
      this.code = this.post.code;
      this.category = this.post.category;
      this.previewPicture = this.post.previewPicture;
      this.previewText = this.post.previewText;
      this.detailText = this.post.detailText;
    },
  },
  validations() {
    const checkTitle = (value) => {
      if (typeof value === 'undefined' || value === null || value === '') {
        return true;
      }
      return !/[^(а-яА-ЯЁёa-zA-Z0-9,.:;!?\-\s)]/.test(value);
    };

    const checkCode = (value) => {
      if (typeof value === 'undefined' || value === null || value === '') {
        return true;
      }
      return !/[^(a-zA-Z0-9)]/.test(value);
    };

    return {
      title: { checkTitle, required },
      category: { checkTitle },
      code: { checkCode, required },
      detailText: { required },
    };
  },
};
</script>

<style lang="stylus" scoped>
  .publish
    card()
    opacity 0
    transition opacity .2s

    &._loaded
      opacity 1

    .title
      margin-bottom 45px

      h1
        font-family 'Roboto Slab', sans-serif
        margin 0
        font-size 34px
        font-weight normal

    .row

      &._controls
        text-align right

        button + button
          margin-left 10px

      & + .row
        margin-top 25px

    .row-title
      font-family 'Roboto Slab', sans-serif
      font-size 18px
      margin-bottom 15px

    .question
      position relative
      top 2px
      left 10px
      cursor pointer

    +phone()
      padding-top 20px

      .title
        margin-bottom 25px
        text-align center

        h1
          font-size 24px

      .row-title
        font-size 16px
        margin-bottom 10px

      .row

        & + .row
          margin-top 15px

        &._controls
          margin-top 20px

</style>
