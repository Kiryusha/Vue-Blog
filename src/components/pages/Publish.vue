<template lang="pug">
  section.publish
    .container
      .publish__content
        Form(
          :postCode="code"
        )
</template>

<script>
import Form from 'Components/elements/publish/Form';

export default {
  components: {
    Form,
  },
  data() {
    return {
      code: null,
    };
  },
  created() {
    const { code } = this.$route.params;

    if (code) {
      this.code = code;
    }
  },
};
</script>

<style lang="stylus">
  .publish
    min-height 100%
    padding-top 60px
    padding-bottom 250px
    background-attachment fixed
    background-image: url("data:image/svg+xml,%3Csvg width='40' height='1' viewBox='0 0 40 1' xmlns='http://www.w3.org/2000/svg'%3E%3Cpath d='M0 0h20v1H0z' fill='%239C92AC' fill-opacity='0.1' fill-rule='evenodd'/%3E%3C/svg%3E")

    +phone()
      padding-top 20px
      padding-bottom 60px

</style>
