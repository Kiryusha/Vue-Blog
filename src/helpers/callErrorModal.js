// calling error modal
export default function (context, error) {
  let errText = 'Неизвестная ошибка.';

  if (error.response) {
    errText = error.response.data.message ? error.response.data.message :
      error.response.data;
  } else {
    errText = error.message;
  }

  context.$Progress.fail();
  context.$modal.show('response', { message: errText });
}
