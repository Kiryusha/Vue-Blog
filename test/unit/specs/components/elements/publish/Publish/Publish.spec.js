// Publish.vue

import { shallowMount, mount, createLocalVue } from '@vue/test-utils';
import Vuex from 'vuex';
import Publish from 'Components/elements/publish/Publish.vue';
import callErrorModal from '@/helpers/callErrorModal';

const localVue = createLocalVue();
localVue.use(Vuex);
localVue.directive('tooltip',  () => {});

jest.mock('@/helpers/callErrorModal', () => jest.fn());

describe('Publish.vue', () => {
  const fetchPost = jest.fn();
  const updatePost = jest.fn();
  const sendPost = jest.fn();
  const fetchData = jest.fn();
  const submit = jest.fn();
  const config = {
    localVue,
    mocks: {
      $router: {
        push: jest.fn(),
      },
      $Progress: {
        finish: () => {},
        fail: () => {},
      },
      $modal: {
        show: jest.fn(),
      },
      $v: {
        $touch: () => {},
        title: {
          $error: () => {},
          $invalid: false,
          $touch: () => {},
        },
        code: {
          $error: () => {},
          $invalid: true,
          $touch: () => {},
        },
        category: {
          $error: () => {},
          $invalid: false,
          $touch: () => {},
        },
        detailText: {
          $error: () => {},
          $invalid: false,
          $touch: () => {},
        },
      },
    },
  };

  let storeConfig = {
    state: {
      user: {
        id: '1a',
        name: 'test',
        post: {},
      },
    },
    actions: {
      fetchPost: jest.fn(),
      sendPost: jest.fn(),
      updatePost: jest.fn(),
    },
  };
  let store = new Vuex.Store(storeConfig);

  afterEach(() => {
    jest.clearAllMocks();
  });

  describe('component shallow rendering', () => {
    it('should match a snapshot - post state', () => {
      const wrapper = shallowMount(Publish, {...config, store});

      expect(wrapper.html()).toMatchSnapshot();
    });

    it('should match a snapshot - edit state', () => {
      const wrapper = shallowMount(Publish, {...config, store, methods: {
        fetchData,
      }});

      wrapper.setProps({ postCode: 'bar' });

      expect(wrapper.html()).toMatchSnapshot();
    });
  });

  describe('lifecycle-driven methods calls', () => {
    it('should call fetchData() on a created hook and edit state', () => {
      const wrapper = shallowMount(Publish, {...config, store, methods: {
        fetchData,
      }});

      wrapper.setProps({ postCode: 'bar' });

      expect(fetchData).toHaveBeenCalledTimes(1);
      expect(fetchData).toHaveBeenCalledWith('bar');
    });

    it('should call error modal on fetchData() if get error', async () => {
      const fetchPost = jest.fn().mockRejectedValue(new Error('test'))
      const wrapper = shallowMount(Publish, {...config, store, methods: {
        fetchPost,
      }});

      await wrapper.setProps({ postCode: 'bar' });
      await wrapper.vm.$nextTick();

      expect(callErrorModal).toHaveBeenCalledTimes(1);
    });
  });

  describe("custom actions", () => {
    it('should call sendPost() on submit() if values are valid and state is post', () => {
      const wrapper = shallowMount(Publish, {...config, store, methods: {
        sendPost,
      }});
      const el = wrapper.find('.submit');

      wrapper.vm.submit();

      expect(sendPost).toHaveBeenCalledTimes(1);
    });

    it('should call updatePost() on submit() if values are valid and state is edit', () => {
      const wrapper = shallowMount(Publish, {...config, store, methods: {
        updatePost,
      }});

      wrapper.setProps({ postCode: 'bar' });
      wrapper.vm.submit();

      expect(updatePost).toHaveBeenCalledTimes(1);
    });

    it('should call modal on submit() if get response', async () => {
      const sendPost = jest.fn().mockResolvedValue({
        data: {
          message: 'test',
        },
      });
      const wrapper = shallowMount(Publish, {...config, store, methods: {
        sendPost,
      }});

      await wrapper.vm.submit();

      expect(wrapper.vm.$modal.show).toHaveBeenCalledTimes(1);
      expect(wrapper.vm.$modal.show).toHaveBeenCalledWith('response', {
        message: 'test'
      });
    });

    it('should call error modal on submit() if get error', async () => {
      const sendPost = jest.fn().mockRejectedValue(new Error('test'))
      const wrapper = shallowMount(Publish, {...config, store, methods: {
        sendPost,
      }});

      await wrapper.vm.submit();

      expect(callErrorModal).toHaveBeenCalledTimes(1);
    });

    // it('should ', async () => {
    //   const sendPost = jest.fn().mockRejectedValue(new Error('test'))
    //   const wrapper = shallowMount(Publish, {...config, store, methods: {
    //     sendPost,
    //   }});
    //
    //   await wrapper.vm.submit();
    //
    //   expect(callErrorModal).toHaveBeenCalledTimes(1);
    // });
  });
});
