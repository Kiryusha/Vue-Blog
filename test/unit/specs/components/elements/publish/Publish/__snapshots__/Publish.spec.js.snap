// Jest Snapshot v1, https://goo.gl/fbAQLP

exports[`Publish.vue component shallow rendering should match a snapshot - edit state 1`] = `
<section class="publish _loaded">
  <div class="title">
    <h1>Отправить новость</h1>
  </div>
  <div class="content">
    <form class="form">
      <div class="row">
        <div class="row-title">Название</div>
        <textarea-stub></textarea-stub>
      </div>
      <div class="row">
        <div class="row-title"><span>Символьный код</span><span class="question"><icon-stub width="16" height="16"></icon-stub></span></div>
        <textarea-stub></textarea-stub>
      </div>
      <div class="row">
        <div class="row-title">Категория</div>
        <textarea-stub></textarea-stub>
      </div>
      <div class="row">
        <div class="row-title">Путь к картинке для анонса</div>
        <textarea-stub></textarea-stub>
      </div>
      <div class="row">
        <div class="row-title">Описание для анонса</div>
        <textarea-stub></textarea-stub>
      </div>
      <div class="row">
        <div class="row-title"><span>Детальное описание</span><span class="question"><icon-stub width="16" height="16"></icon-stub></span></div>
        <textarea-stub></textarea-stub>
      </div>
      <div class="row _controls">
        <button-stub class="submit"></button-stub>
        <button-stub class="cancel"></button-stub>
      </div>
    </form>
  </div>
</section>
`;

exports[`Publish.vue component shallow rendering should match a snapshot - post state 1`] = `
<section class="publish _loaded">
  <div class="title">
    <h1>Отправить новость</h1>
  </div>
  <div class="content">
    <form class="form">
      <div class="row">
        <div class="row-title">Название</div>
        <textarea-stub></textarea-stub>
      </div>
      <div class="row">
        <div class="row-title"><span>Символьный код</span><span class="question"><icon-stub width="16" height="16"></icon-stub></span></div>
        <textarea-stub></textarea-stub>
      </div>
      <div class="row">
        <div class="row-title">Категория</div>
        <textarea-stub></textarea-stub>
      </div>
      <div class="row">
        <div class="row-title">Путь к картинке для анонса</div>
        <textarea-stub></textarea-stub>
      </div>
      <div class="row">
        <div class="row-title">Описание для анонса</div>
        <textarea-stub></textarea-stub>
      </div>
      <div class="row">
        <div class="row-title"><span>Детальное описание</span><span class="question"><icon-stub width="16" height="16"></icon-stub></span></div>
        <textarea-stub></textarea-stub>
      </div>
      <div class="row _controls">
        <button-stub class="submit"></button-stub>
        <!---->
      </div>
    </form>
  </div>
</section>
`;
